src/index.js

```diff
@@ -607,6 +607,96 @@ export default function (babel) {
     path.unshiftContainer("body", declarations);
   }
 
+  function generateForInIterator (path, type: "array" | "object") {
+    const idx = path.node.idx || path.scope.generateUidIdentifier("i");
+    const len = path.scope.generateUidIdentifier("len");
+
+    const initDeclarations = [
+      t.variableDeclarator(idx, t.numericLiteral(0))
+    ];
+
+    let refId;
+    if (path.get(type).isIdentifier()) {
+      refId = path.node[type];
+    } else {
+      // if the target of iteration is a complex expression,
+      // create a reference so it only evaluates once
+      refId = path.scope.generateUidIdentifier(type.slice(0, 3));
+      initDeclarations.unshift(
+        t.variableDeclarator(
+          refId,
+          path.node[type]
+        )
+      );
+    }
+
+    let keys;
+    if (type === "object") {
+      keys = path.scope.generateUidIdentifier("keys");
+      initDeclarations.push(
+        t.variableDeclarator(keys,
+          t.callExpression(
+            t.memberExpression(
+              t.identifier("Object"),
+              t.identifier("keys")),
+            [refId]
+          )
+        )
+      );
+    }
+
+    initDeclarations.push(
+      t.variableDeclarator(
+        len,
+        t.memberExpression(
+          type === "object" ? keys : refId,
+          t.identifier("length")
+        )
+      )
+    );
+
+    const init = t.variableDeclaration("let", initDeclarations);
+    // _i < _len
+    const test = t.binaryExpression("<", idx, len);
+    // _i++
+    const update = t.updateExpression("++", idx);
+
+    ensureBlockBody(path);
+    const innerDeclarations = [];
+    if (type === "object") {
+      const key = path.node.key || path.scope.generateUidIdentifier("k");
+      innerDeclarations.push(
+        t.variableDeclarator(key, t.memberExpression(keys, idx, true))
+      );
+
+      if (path.node.val) {
+        innerDeclarations.push(
+          t.variableDeclarator(
+            path.node.val,
+            t.memberExpression(refId, key, true)
+          )
+        );
+      }
+
+      const declarations = t.variableDeclaration("const", innerDeclarations);
+      path.get("body").unshiftContainer("body", declarations);
+    } else {
+      if (path.node.elem) {
+        innerDeclarations.push(
+          t.variableDeclarator(
+            path.node.elem,
+            t.memberExpression(refId, idx, true)
+          )
+        );
+
+        const declarations = t.variableDeclaration("const", innerDeclarations);
+        path.get("body").unshiftContainer("body", declarations);
```

Comment:
Ah yes I can do that. I have it done a different way currently but it's a little more naive.